src/io/fasta.rs

```diff
@@ -252,12 +361,73 @@ struct IndexRecord {
 
 
 /// A sequence record returned by the FASTA index.
+#[derive(Debug, PartialEq)]
 pub struct Sequence {
     pub name: String,
     pub len: u64,
 }
 
 
+pub struct IndexedReaderIterator<'a, R: io::Read + io::Seek + 'a> {
+    reader: &'a mut IndexedReader<R>,
+    record: IndexRecord,
+    bases_left: u64,
+    line_offset: u64,
+    buf: Vec<u8>,
+    buf_idx: usize,
+    buf_len: usize,
+}
+
+
+impl<'a, R: io::Read + io::Seek + 'a> IndexedReaderIterator<'a, R> {
+    fn fill_buffer(&mut self) -> io::Result<()> {
+        assert!(self.bases_left > 0);
+
+        while self.buf_idx >= self.buf_len {
+            let bases_read = self.reader
+                .read_line(&self.record,
+                           &mut self.line_offset,
+                           self.bases_left,
+                           &mut self.buf)?;
+
+            self.buf_idx = 0;
+            self.buf_len = bases_read as usize;
+            self.bases_left -= bases_read;
+        }
+
+        Ok(())
+    }
+}
+
+
+impl<'a, R: io::Read + io::Seek + 'a> Iterator for IndexedReaderIterator<'a, R> {
+    type Item = io::Result<u8>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.buf_idx < self.buf_len {
+            let item = Some(Ok(self.buf[self.buf_idx]));
```

Comment:
To expand on the previous, I ran some additional benchmarks to see what the cost of the iterator was in practice, for two block sizes and with or without the file cached, on a SSD and a HDD. The TL;DR is there seems to be an overhead of about 20-40% when the file is cached (hot), but no negative change in runtime when the file is uncached (cold). Of course, cache or no cache makes a much bigger difference than the choice of read or iter. I also seem to have mentally mixed up the cost from blocked reads on long lines with the seek overhead, so please disregard the comments I made with regards to seeks. The results are here:

https://gist.github.com/MikkelSchubert/ea5a70776da166aa10e8f1f5efce57ba

However, should you not want these iterators as part of hts-bio, then I can easily break out the related improvements and make a new pull request with just those (i.e. the fix for with_fasta_file, capping memory usage, and additional tests).